### ISSUE TITLE
Improve Buf / BufMut docs

### ISSUE THREAD

**Issue by carllerche** [carllerche/bytes#178]
It is not clear how these traits should be implemented when the byte storage is not in continuous memory.

**Comment by StanislavGlebik** [comment]
Yeah, we just run into an issue. For our datastructure that's not in continuous memory we implemented `Buf` trait. Each of the continuous pieces of this datastructure can be empty, and so `Buf::bytes()` might be empty. But if it happens that `get_u8()` is called when `Buf::bytes()` returns an empty slice then it panics i.e. `get_u8()` doesn't check if a slice is empty or not

**Comment by carllerche** [comment]
Do you have thoughts on how to improve the docs? I would :heart: a PR.
